src/librustc_mir/interpret/eval_context.rs

```diff
@@ -223,12 +223,17 @@ pub(super) fn mir_assign_valid_types<'tcx>(
     // Type-changing assignments can happen for (at least) two reasons:
     // - `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
     // - Subtyping is used. While all normal lifetimes are erased, higher-ranked lifetime
-    //   bounds are still around and can lead to type differences.
-    // There is no good way to check the latter, so we compare layouts instead -- but only
-    // for values with `Scalar`/`ScalarPair` abi.
-    // FIXME: Do something more accurate, type-based.
-    match &src.abi {
-        Abi::Scalar(..) | Abi::ScalarPair(..) => src.layout == dest.layout,
+    //   bounds on function pointers are still around and can lead to type differences.
+    match (&src.ty.kind, &dest.ty.kind) {
+        (ty::Ref(_, src_pointee, _), ty::Ref(_, dest_pointee, _)) => {
+            // After optimizations, there can be assignments that change reference mutability.
+            // This does not affect reference layout, so that is fine.
+            src_pointee == dest_pointee
```

Comment:
I am slightly paranoid here about comparing the wrong thing (in an earlier version I accidentally compared mutabilities here instead of types, by using the wrong pattern). Is there any good way to assert that these things are types?

Reply:
You can probably use `Ty::eq` to call the implementation of `==`, but idk if it looks pretty.
The other option is I guess ascription, i.e. `(src_pointeee: Ty) == (dest_pointee: Ty)`.